### PR TITLE
Bulk code generation

### LINKED ISSUE #1: Generate copy method for value objects
Given a value object defined as a `struct` in Swift, generate a copy method; for instance:

```swift
struct Article {
  let title: String
  let subtitle: String?
  let tags: [String]
}
```

We would generate:

```swift
extension Article {
  func copy(withTitle title: String? = nil,
            withSubtitle subtitle: String?? = nil,
            withTags tags: [String]? = nil) -> Article {
    Article(title: title ?? self.title, 
            subtitle: subtitle ?? self.subtitle, 
            tags: tags ?? self.tags)
  }
}
```

### PR DESCRIPTION
**Pull request by truizlop**
## Related issues

- Close #1 

## Goal

Generate code for multiple files based on a provided visitor. Use this code to generate the copy visitor in multiple files.

## Implementation details

This PR provides algebras for code generation and accessing the file system that have production implementations based on the macOS file system and a code generator based on Swift Syntax that receives a `CodegenVisitor` to do this job.

## Testing details

Snapshot testing is used to verify the output of the generation process is correct. A snapshotting strategy is created to check line by line the output of the generation process.

## Trying it out

You can try it out in a new project. You need to follow these steps:

- Add this project as a dependency to your project in Swift Package Manager
- Create a new `Aggregate` target.
- In its `Build Phases` step, add a new `Run script` step.
- Paste the following code:

```sh
cd ${BUILD_DIR}/../../SourcePackages/checkouts/bow-meta
swift run generate-optics <input> <output>
```

where:
- `<input>` is the root directory of all files that the script needs to parse and generate.
- `<output>` is the directory where the generated code will be saved.

You can provide a relative path from your project by using the `${SRCROOT}` variable, for instance:

```
cd ${BUILD_DIR}/../../SourcePackages/checkouts/bow-meta
swift run generate-optics ${SRCROOT}/Blog/Models ${SRCROOT}/Blog/Output
```

*Note:* notice that we have to change directory based on `${BUILD_DIR}` since there is no variable at the moment to refer to the checkouts directory. Perhaps that will change in the future. We can check all defined variables by running:

```sh
xcodebuild -project myProj.xcodeproj -target "myTarg" -showBuildSettings
```

### CODE REVIEW

File: Tests/MetaTests/Extensions/Snapshotting+Codegen.swift [hunk 1]
```diff
@@ -15,4 +16,14 @@ extension Snapshotting where Value == URL, Format == String {
     }
     
     static let copyMethod: Snapshotting<URL, String> = .generatedCode(CopyVisitor())
```
Comment (miguelangel-dev) [inline Tests/MetaTests/Extensions/Snapshotting+Codegen.swift]: you should delete this hardcoded `generatedCode` and use instead `generatedCode<T: CodegenVisitor>(_ visitor: T)` where you need it

File: Tests/MetaTests/Extensions/Snapshotting+Codegen.swift [hunk 2]
```diff
@@ -15,4 +16,14 @@ extension Snapshotting where Value == URL, Format == String {
...
+    
+    static func copyGeneration<D: CodegenDependencies>(for environment: D) -> Snapshotting<URL, String> {
```
Comment (miguelangel-dev) [inline Tests/MetaTests/Extensions/Snapshotting+Codegen.swift]: IMO both methods are the same, one working with `visitor` and the other using `environment` but should be named `generatedCode`, aren't they?

**Review by miguelangel-dev — APPROVED** [review]
Good work ! :)

File: Tests/MetaTests/Extensions/Snapshotting+Codegen.swift (hunk 1 above)
Comment (truizlop) [reply Tests/MetaTests/Extensions/Snapshotting+Codegen.swift]: It is just a utility for brevity, I think it is nice to have more readable tests, even though it relies in the parametric one.

File: Tests/MetaTests/Extensions/Snapshotting+Codegen.swift (hunk 2 above)
Comment (truizlop) [reply Tests/MetaTests/Extensions/Snapshotting+Codegen.swift]: They are not the same, the one for the visitor only does the check for a single file generation and the other one receives a directory and recursively checks and generates all files. Naming these two different things is a bit complex, but I am open to renaming it to clarify.

**Pull request merged**